README.md
# fluent-plugin-measure_time

[![Build Status](https://secure.travis-ci.org/sonots/fluent-plugin-measure_time.png?branch=master)](http://travis-ci.org/sonots/fluent-plugin-measure_time)
[![Code Climate](https://codeclimate.com/github/sonots/fluent-plugin-measure_time.png)](https://codeclimate.com/github/sonots/fluent-plugin-measure_time)

Fluentd plugin to measure elapsed time to process messages


## Installation

Use RubyGems:

    gem install fluent-plugin-measure_time

## Parameters

* tag

    The output tag name. Default is `measure_time`

* hook (required)

    Specify the method to measure time.

* interval

    The time interval to emit measurement results. Default is nil which do not compute statistics and emit the time in each measurement.

## Configuration Example 1 - Profile an Output Plugin

As an example, let's profile how long the [emit](https://github.com/sonots/fluent-plugin-grep/blob/master/lib/fluent/plugin/out_grep.rb#L56) method of [fluent-plugin-grep](https://github.com/sonots/fluent-plugin-grep) is taking.
Configure fluentd.conf as below:

For Fluentd v0.10:

```apache
<source>
  type measure_time
  # This makes available the `measure_time` directive for all plugins
</source>

<source>
  type forward
  port 24224
</source>

# measure_time plugin output comes here
<match measure_time>
  type stdout
</match>

# Whatever you want to do
<match greped.**>
  type stdout
</match>

<match **>
  type grep
  add_tag_prefix greped
  <measure_time>
    tag measure_time
    hook emit
  </measure_time>
</match>
```

For Fluentd v0.12:

```apache
<label @measure_time>
  <match>
    @type measure_time
    # This makes available the `measure_time` directive for all plugins
  </match>
</label>

<source>
  @type dummy
  tag raw.dummy
  dummy {"message":"foo"}
</source>

# measure_time plugin output comes here
<match measure_time>
  @type stdout
</match>

# Whatever you want to do
<match greped.**>
  @type stdout
</match>

<match **>
  @type grep
  add_tag_prefix greped
  <measure_time>
    tag measure_time
    hook emit
  </measure_time>
</match>
```

The output of fluent-plugin-measure_time will be as below:

```
measure_time: {"time":0.000849735,"class":"Fluent::GrepOutput","hook":"emit","object_id":83935080}
```

where `time` denotes the measured elapsed time, and `class`, `hook`, and `object_id` denotes the hooked class, the hooked method, and the object id of the plugin instance.

### interval option

fluent-plugin-measure_time outputs the elapsed time for each calling, but you can use the `interval` option when you want to get statistics in each interval.

```
measure_time: {"max":1.011,"avg":0.002","num":10,"class":"Fluent::GrepOutput","hook":"emit","object_id":83935080}
```

where `max` and `avg` are the maximum and average elapsed times, and `num` is the number of being called in each interval.

## Configuration Example (2) - Profile the in_forward plugin

I introduce an interesting example here.

Following illustration draws the sequence of that `in_forward` plugin receives a data, processes, and passes the data to output plugins.

*Sequence Diagram*

```
     +–––––––––––––+    +––––––––––––––+   +––––––––––––––+
     |  in_forwrd  |    |   Output     |   |   Output     |
     +––––––+––––––+    +––––––+–––––––+   +––––––+–––––––+
#on_message | start = Time.now |                  |
            +––––––––––––––––––>                  |
            |      #emit       |                  |
            |                  +––––––––––––––––––>
            |                  |      #emit       |
            |                  |                  |
            |                  |                  |
            |                  <– – – – – – – – – +
            | elapsed = Time.now - start          |
            <– – – – – - – – – +                  |
            |                  |                  |
            +                  +                  +
```

As the illustration, by hooking `on_message` method of `in_forward` plugin,
we can measure the blocking time taking to process the received data,
which also means that the time taking until `in_forward` will be ready for receiving a next data.

This profiling is very useful to investigate when you have a suspicion that throughputs of Fluentd fell down extremely.

The configuration will be as follows:

For Fluentd v0.10:

```apache
<source>
  type measure_time
  # This makes available the `measure_time` directive for all plugins
</source>

<source>
  type forward
  port 24224
  <measure_time>
    tag measure_time
    hook on_message
  </measure_time>
</source>

<match measure_time>
  type stdout
</match>

# whatever you want
<match **>
  type stdout
</match>
```

For Fluentd v0.12:

```apache
<label @measure_time>
  <match>
    @type measure_time
    # This makes available the `measure_time` directive for all plugins
  </match>
</match>

<source>
  @type forward
  port 24224
  <measure_time>
    tag measure_time
    hook on_message
  </measure_time>
</source>

<match measure_time>
  @type stdout
</match>

# whatever you want
<match **>
  @type stdout
</match>
```

Output becomes as below:

```
measure_time: {"time":0.000849735,"class":"Fluent::ForwardInput","hook":"on_message","object_id":83935080}
```

## ChangeLog

See [CHANGELOG.md](CHANGELOG.md) for details.

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new [Pull Request](../../pull/new/master)

## Copyright

Copyright (c) 2014 Naotoshi Seo. See [LICENSE](LICENSE) for details.


CHANGELOG.md
## 0.2.1  (2016/04/09)

Enhancements:

* raise ConfigError to use `<label @measure_time><match></match></label>` instead of `<source></source>` for Fluentd v0.12.

## 0.2.0  (2016/04/09)

Enhancements:

* Make it work with Fluentd v0.12

## 0.1.2  (2014/09/22)

Changes:

* Remove fluentd version constraint

## 0.1.1  (2014/04/12)

Fixes:

* Fix emit messages

## 0.1.0  (2014/04/12)

First release



spec/spec_helper.rb
# encoding: UTF-8
require 'rubygems'
require 'bundler'
require 'fluent/load'
Bundler.setup(:default, :test)
Bundler.require(:default, :test)

require 'fluent/test'
require 'rspec'
require 'rspec/its'
require 'pry'

$TESTING=true
$:.unshift File.join(File.dirname(__FILE__), '..', 'lib')
require 'fluent/plugin/in_measure_time'
require 'fluent/plugin/out_measure_time'


spec/elapsed_time_spec.rb
# encoding: UTF-8
require_relative 'spec_helper'
require 'fluent/plugin/in_forward'
require 'fluent/plugin/out_stdout'

describe Fluent::ForwardInput do
  before { Fluent::Test.setup }

  def create_driver(conf=CONFIG)
    Fluent::Test::InputTestDriver.new(Fluent::ForwardInput).configure(conf)
  end

  def connect
    TCPSocket.new('127.0.0.1', PORT)
  end

  def send_data(data)
    io = connect
    begin
      io.write data
    ensure
      io.close
    end
  end

  def self.unused_port
    s = TCPServer.open(0)
    port = s.addr[1]
    s.close
    port
  end

  PORT = unused_port
  CONFIG = %[
    port #{PORT}
    bind 127.0.0.1
  ]

  let(:driver) { create_driver(config) }

  describe 'test configure' do
    let(:config) {CONFIG + %[
      <elapsed>
        tag test
        interval 10
        hook on_message
      </elapsed>
    ]}
    let(:subject) { driver.instance.elapsed }
    its(:tag) { should == 'test' }
    its(:interval) { should == 10 }
    its(:hook) { should == 'on_message' }
  end

  describe 'test emit' do
    let(:config) {CONFIG + %[
      <elapsed>
        tag elapsed
        interval 1
        # hook Fluent::ForwardInput::Handler.on_read # not support inner class yet
        hook Fluent::ForwardInput.on_message
      </elapsed>
    ]}
    it 'should flush' do
      d = driver.instance
      d.__send__(:on_message, ['tag1', 0, {'a'=>1}].to_msgpack)
      triple = d.elapsed.flush(0)
      triple[0].should == 'elapsed'
      triple[2].keys.should =~ [:num, :max, :avg]
    end
  end
end

describe Fluent::StdoutOutput do
  before { Fluent::Test.setup }

  def create_driver(conf=CONFIG, tag = 'test')
    Fluent::Test::OutputTestDriver.new(Fluent::StdoutOutput, tag).configure(conf)
  end

  CONFIG = %[
  ]

  let(:driver) { create_driver(config) }

  describe 'test configure' do
    let(:config) {CONFIG + %[
      <elapsed>
        tag test
        interval 10
        hook emit
      </elapsed>
    ]}
    let(:subject) { driver.instance.instance_variable_get(:@elapsed) }
    its(:tag) { should == 'test' }
    its(:interval) { should == 10 }
    its(:hook) { should == 'emit' }
  end

  describe 'test emit' do
    let(:config) {CONFIG + %[
      <elapsed>
        tag elapsed
        interval 1
        hook emit
      </elapsed>
    ]}
    it 'should flush' do
      d = driver.instance
      d.emit('tag1', Fluent::OneEventStream.new(0, {'a'=>1}), Fluent::NullOutputChain.instance)
      triple = d.elapsed.flush(0)
      triple[0].should == 'elapsed'
      triple[2].keys.should =~ [:num, :max, :avg]
    end
  end
end



spec/in_measure_time_spec.rb
# encoding: UTF-8
require_relative 'spec_helper'
require 'fluent/plugin/in_forward'
require 'fluent/plugin/out_stdout'

describe Fluent::MeasureTimeInput do
  before { Fluent::Test.setup }

  def create_driver(conf=%[])
    d = Fluent::Test::InputTestDriver.new(Fluent::MeasureTimeInput).configure(conf)
    unless d.respond_to?(:router)
      d.singleton_class.send(:define_method, :router) { ::Fluent::Engine }
    end
  end

  describe 'test configure' do
    it { expect { create_driver }.not_to raise_error }
  end
end

describe "extends Fluent::ForwardInput" do
  before { Fluent::Test.setup }

  def create_driver(conf=CONFIG)
    Fluent::MeasureTimeInput.new.configure("")
    Fluent::Test::InputTestDriver.new(Fluent::ForwardInput).configure(conf)
  end

  def connect
    TCPSocket.new('127.0.0.1', PORT)
  end

  def send_data(data)
    io = connect
    begin
      io.write data
    ensure
      io.close
    end
  end

  def self.unused_port
    s = TCPServer.open(0)
    port = s.addr[1]
    s.close
    port
  end

  PORT = unused_port
  CONFIG = %[
    port #{PORT}
    bind 127.0.0.1
  ]

  let(:driver) { create_driver(config) }

  describe 'test configure' do
    let(:config) {CONFIG + %[
      <measure_time>
        tag test
        interval 10
        hook on_message
      </measure_time>
    ]}
    let(:subject) { driver.instance.measure_time }
    its(:tag) { should == 'test' }
    its(:interval) { should == 10 }
    its(:hook) { should == 'on_message' }
  end

  describe 'test emit' do
    let(:config) {CONFIG + %[
      <measure_time>
        tag measure_time
        interval 1
        hook on_message
      </measure_time>
    ]}
    it 'should flush' do
      d = driver.instance
      data = ['tag1', 0, {'a'=>1}].to_msgpack
      d.__send__(:on_message, data, data.bytesize, "hi, yay!")
      triple = d.measure_time.flush(0)
      expect(triple[0]).to eql('measure_time')
      expect(triple[2].keys).to eql([:max, :avg, :num, :class, :hook, :object_id])
    end
  end
end

describe "extends Fluent::StdoutOutput" do
  before { Fluent::Test.setup }

  def create_driver(conf=CONFIG, tag = 'test')
    Fluent::Test::OutputTestDriver.new(Fluent::StdoutOutput, tag).configure(conf)
  end

  CONFIG = %[
  ]

  let(:driver) { create_driver(config) }

  describe 'test configure' do
    let(:config) {CONFIG + %[
      <measure_time>
        tag test
        interval 10
        hook emit
      </measure_time>
    ]}
    let(:subject) { driver.instance.measure_time }
    its(:tag) { should == 'test' }
    its(:interval) { should == 10 }
    its(:hook) { should == 'emit' }
  end

  describe 'test emit' do
    let(:config) {CONFIG + %[
      <measure_time>
        tag measure_time
        hook emit
      </measure_time>
    ]}
    it 'should flush' do
      time = Fluent::Engine.now
      allow(Fluent::Engine).to receive(:now) { time }
      d = driver.instance
      expect(d.router).to receive(:emit) # .with("measure_time", time, {})
      d.emit('tag1', Fluent::OneEventStream.new(0, {'a'=>1}), Fluent::NullOutputChain.instance)
    end
  end

  describe 'test interval' do
    let(:config) {CONFIG + %[
      <measure_time>
        tag measure_time
        interval 1
        hook emit
      </measure_time>
    ]}
    it 'should flush' do
      d = driver.instance
      d.emit('tag1', Fluent::OneEventStream.new(0, {'a'=>1}), Fluent::NullOutputChain.instance)
      triple = d.measure_time.flush(0)
      expect(triple[0]).to eql('measure_time')
      expect(triple[2].keys).to eql([:max, :avg, :num, :class, :hook, :object_id])
    end
  end
end



lib/fluent/mixin/elapsed_time.rb
require 'fluent/input'

module Fluent
  module ElapsedTimable
    def self.included(klass)
      klass.__send__(:alias_method, :configure_without_elapsed, :configure)
      klass.__send__(:alias_method, :configure, :configure_with_elapsed)
    end

    def configure_with_elapsed(conf)
      configure_without_elapsed(conf)
      if element = conf.elements.select { |element| element.name == 'elapsed' }.first
        @elapsed = ElapsedTime.new(self, log)
        @elapsed.configure(element)
        # #start and #stop methods must be extended in concrete input plugins
        # because most of built-in input plugins do not call `super`
        klass = self.class
        unless klass.method_defined?(:start_without_elapsed)
          klass.__send__(:alias_method, :start_without_elapsed, :start)
          klass.__send__(:alias_method, :start, :start_with_elapsed)
          klass.__send__(:alias_method, :shutdown_without_elapsed, :shutdown)
          klass.__send__(:alias_method, :shutdown, :shutdown_with_elapsed)
        end
      end
    end

    def elapsed
      @elapsed
    end

    def start_with_elapsed
      start_without_elapsed
      @elapsed.start if @elapsed
    end

    def shutdown_with_elapsed
      shutdown_without_elapsed
      @elapsed.stop if @elapsed
    end
  end

  Input.__send__(:include, ElapsedTimable)
  Output.__send__(:include, ElapsedTimable)

  class ElapsedTime
    attr_reader :input, :log, :times, :mutex, :thread, :tag, :interval, :hook
    def initialize(input, log)
      @input = input
      @log = log
      @times = []
      @mutex = Mutex.new
    end

    def configure(conf)
      @tag = conf['tag'] || 'elapsed'
      @interval = conf['interval'].to_i || 60
      unless @hook = conf['hook']
        raise Fluent::ConfigError, '`hook` option must be specified in <elapsed></elpased> directive'
      end
      apply_hook(@hook)
    end

    def add(time)
      @times << time
    end

    def clear
      @times.clear
    end

    def apply_hook(hook)
      if hook.include?('.')
        klass_name, method_name = hook.split('.', 2)
        klass = constantize(klass_name)
      else
        klass = @input.class
        method_name = hook
      end
      old_method_name = "#{method_name}_without_elapsed".to_sym 
      unless klass.method_defined?(old_method_name)
        klass.__send__(:alias_method, old_method_name, method_name)
        klass.__send__(:define_method, method_name) do |*args|
          elapsed.measure_time(klass, method_name) do
            self.__send__(old_method_name, *args)
          end
        end
      end
    end

    def measure_time(klass, method_name)
      started = Time.now
      output = yield
      elapsed = (Time.now - started).to_f
      log.info "elapsed time at #{klass}##{method_name} is #{elapsed} sec"
      @mutex.synchronize { self.add(elapsed) }
      output
    end

    def start
      @thread = Thread.new(&method(:run))
    end

    def stop
      @thread.terminate
      @thread.join
    end

    def run
      @last_checked ||= Engine.now
      while (sleep 0.5)
        begin
          now = Engine.now
          if now - @last_checked >= @interval
            flush(now)
            @last_checked = now
          end
        rescue => e
          log.warn "in_forward: #{e.class} #{e.message} #{e.backtrace.first}"
        end
      end
    end

    def flush(now)
      times = []
      @mutex.synchronize do
        times = @times.dup
        self.clear
      end
      triple = nil
      unless times.empty?
        num = times.size
        max = num == 0 ? 0 : times.max
        avg = num == 0 ? 0 : times.map(&:to_f).inject(:+) / num.to_f
        triple = [@tag, now, {:num => num, :max => max, :avg => avg}]
        Engine.emit(*triple)
      end
      triple
    end

    private
    # File activesupport/lib/active_support/inflector/methods.rb, line 219
    def constantize(camel_cased_word)
      names = camel_cased_word.split('::')
      names.shift if names.empty? || names.first.empty?

      names.inject(Object) do |constant, name|
        if constant == Object
          constant.const_get(name)
        else
          candidate = constant.const_get(name)
          next candidate if constant.const_defined?(name, false)
          next candidate unless Object.const_defined?(name)

          # Go down the ancestors to check it it's owned
          # directly before we reach Object or the end of ancestors.
          constant = constant.ancestors.inject do |const, ancestor|
            break const    if ancestor == Object
            break ancestor if ancestor.const_defined?(name, false)
            const
          end

          # owner is in Object, so raise
          constant.const_get(name, false)
        end
      end
    end
  end
end


lib/fluent/plugin/measure_timable.rb
module Fluent
  module MeasureTimable
    def self.included(klass)
      unless klass.method_defined?(:configure_without_measure_time)
        klass.__send__(:alias_method, :configure_without_measure_time, :configure)
        klass.__send__(:alias_method, :configure, :configure_with_measure_time)
      end

      unless klass.method_defined?(:router)
        define_method(:router) { ::Fluent::Engine }
      end
    end

    attr_reader :measure_time

    def configure_with_measure_time(conf)
      configure_without_measure_time(conf)
      if element = conf.elements.select { |element| element.name == 'measure_time' }.first
        @measure_time = MeasureTime.new(self, log, router)
        @measure_time.configure(element)
      end
    end
  end

  class MeasureTime
    attr_reader :plugin, :log, :router, :times, :mutex, :thread, :tag, :interval, :hook
    def initialize(plugin, log, router)
      @plugin = plugin
      @klass = @plugin.class
      @log = log
      @router = router
      @times = []
      @mutex = Mutex.new
    end

    def configure(conf)
      @tag = conf['tag'] || 'measure_time'
      unless @hook = conf['hook']
        raise Fluent::ConfigError, '`hook` option must be specified in <measure_time></measure_time> directive'
      end
      @hook_msg = {:class => @klass.to_s, :hook => @hook.to_s, :object_id => @plugin.object_id.to_s}
      @interval = conf['interval'].to_i if conf['interval']
      @add_or_emit_proc =
        if @interval
          # add to calculate statistics in each interval
          Proc.new {|elapsed|
            @mutex.synchronize { @times << elapsed }
          }
        else
          # emit information immediately
          Proc.new {|elapsed|
            msg = {:time => elapsed}.merge(@hook_msg)
            router.emit(@tag, ::Fluent::Engine.now, msg)
          }
        end
      apply_hook
    end

    def apply_hook
      @plugin.instance_eval <<EOF
        def #{@hook}(*args)
          measure_time.measure_time do
            super
          end
        end
        def start
          super
          measure_time.start
        end
        def stop
          super
          measure_time.stop
        end
EOF
    end

    def measure_time
      started = Time.now
      output = yield
      elapsed = (Time.now - started).to_f
      log.debug "elapsed time at #{@klass}##{@hook} is #{elapsed} sec"
      @add_or_emit_proc.call(elapsed)
      output
    end

    def start
      return unless @interval
      @thread = Thread.new(&method(:run))
    end

    def stop
      return unless @interval
      @thread.terminate
      @thread.join
    end

    def run
      @last_checked ||= ::Fluent::Engine.now
      while (sleep 0.5)
        begin
          now = ::Fluent::Engine.now
          if now - @last_checked >= @interval
            flush(now)
            @last_checked = now
          end
        rescue => e
          log.warn "in_measure_time: hook #{@klass}##{@hook} #{e.class} #{e.message} #{e.backtrace.first}"
        end
      end
    end

    def flush(now)
      times = []
      @mutex.synchronize do
        times = @times.dup
        @times.clear
      end
      triple = nil
      unless times.empty?
        num = times.size
        max = num == 0 ? 0 : times.max
        avg = num == 0 ? 0 : times.map(&:to_f).inject(:+) / num.to_f
        triple = [@tag, now, {:max => max, :avg => avg, :num => num}.merge(@hook_msg)]
        router.emit(*triple)
      end
      triple
    end
  end
end


benchmark/plugin/in_forward.rb
#
# Fluent
#
# Copyright (C) 2011 FURUHASHI Sadayuki
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
#
module Fluent


  class ForwardInput < Input
    Plugin.register_input('forward', self)

    def initialize
      super
      require 'fluent/plugin/socket_util'
    end

    config_param :port, :integer, :default => DEFAULT_LISTEN_PORT
    config_param :bind, :string, :default => '0.0.0.0'
    config_param :backlog, :integer, :default => nil
    # SO_LINGER 0 to send RST rather than FIN to avoid lots of connections sitting in TIME_WAIT at src
    config_param :linger_timeout, :integer, :default => 0
    attr_reader :elapsed # for test

    def configure(conf)
      super

      if element = conf.elements.first { |element| element.name == 'elapsed' }
        tag = element["tag"] || 'elapsed'
        interval = element["interval"].to_i || 60
        hook = element['hook'] || 'on_message'
        @elapsed = ElapsedMeasure.new(log, tag, interval, hook)
      end
    end

    def start
      @loop = Coolio::Loop.new

      @lsock = listen
      @loop.attach(@lsock)

      @usock = SocketUtil.create_udp_socket(@bind)
      @usock.bind(@bind, @port)
      @usock.fcntl(Fcntl::F_SETFL, Fcntl::O_NONBLOCK)
      @hbr = HeartbeatRequestHandler.new(@usock, method(:on_heartbeat_request))
      @loop.attach(@hbr)

      @thread = Thread.new(&method(:run))
      @elapsed.start if @elapsed
      @cached_unpacker = $use_msgpack_5 ? nil : MessagePack::Unpacker.new
    end

    def shutdown
      @loop.watchers.each {|w| w.detach }
      @loop.stop
      @usock.close
      listen_address = (@bind == '0.0.0.0' ? '127.0.0.1' : @bind)
      # This line is for connecting listen socket to stop the event loop.
      # We should use more better approach, e.g. using pipe, fixing cool.io with timeout, etc.
      TCPSocket.open(listen_address, @port) {|sock| } # FIXME @thread.join blocks without this line
      @thread.join
      @lsock.close
      @elapsed.stop if @elapsed
    end

    class ElapsedMeasure
      attr_reader :tag, :interval, :hook, :times, :sizes, :mutex, :thread, :log
      def initialize(log, tag, interval, hook)
        @log = log
        @tag = tag
        @interval = interval
        @hook = hook.split(',')
        @times = []
        @sizes = []
        @mutex = Mutex.new
      end

      def add(time, size)
        @times << time
        @sizes << size
      end

      def clear
        @times.clear
        @sizes.clear
      end

      def hookable?(caller)
        @hook.include?(caller.to_s)
      end

      def measure_time(caller, size)
        if hookable?(caller)
          started = Time.now
          yield
          elapsed = (Time.now - started).to_f
          log.debug "in_forward: elapsed time at #{caller} is #{elapsed} sec for #{size} bytes"
          @mutex.synchronize { self.add(elapsed, size) }
        else
          yield
        end
      end

      def start
        @thread = Thread.new(&method(:run))
      end

      def stop
        @thread.terminate
        @thread.join
      end

      def run
        @last_checked ||= Engine.now
        while (sleep 0.5)
          begin
            now = Engine.now
            if now - @last_checked >= @interval
              flush(now)
              @last_checked = now
            end
          rescue => e
            log.warn "in_forward: #{e.class} #{e.message} #{e.backtrace.first}"
          end
        end
      end

      def flush(now)
        times, sizes = [], []
        @mutex.synchronize do
          times = @times.dup
          sizes = @sizes.dup
          self.clear
        end
        if !times.empty? and !sizes.empty?
          num = times.size
          max = num == 0 ? 0 : times.max
          avg = num == 0 ? 0 : times.map(&:to_f).inject(:+) / num.to_f
          size_max = num == 0 ? 0 : sizes.max
          size_avg = num == 0 ? 0 : sizes.map(&:to_f).inject(:+) / num.to_f
          Engine.emit(@tag, now, {:num => num, :max => max, :avg => avg, :size_max => size_max, :size_avg => size_avg})
        end
      end
    end

    def listen
      log.info "listening fluent socket on #{@bind}:#{@port}"
      s = Coolio::TCPServer.new(@bind, @port, Handler, @linger_timeout, log, method(:on_message), @elapsed)
      s.listen(@backlog) unless @backlog.nil?
      s
    end

    #config_param :path, :string, :default => DEFAULT_SOCKET_PATH
    #def listen
    #  if File.exist?(@path)
    #    File.unlink(@path)
    #  end
    #  FileUtils.mkdir_p File.dirname(@path)
    #  log.debug "listening fluent socket on #{@path}"
    #  Coolio::UNIXServer.new(@path, Handler, method(:on_message))
    #end

    def run
      @loop.run
    rescue => e
      log.error "unexpected error", :error => e, :error_class => e.class
      log.error_backtrace
    end

    protected
    # message Entry {
    #   1: long time
    #   2: object record
    # }
    #
    # message Forward {
    #   1: string tag
    #   2: list<Entry> entries
    # }
    #
    # message PackedForward {
    #   1: string tag
    #   2: raw entries  # msgpack stream of Entry
    # }
    #
    # message Message {
    #   1: string tag
    #   2: long? time
    #   3: object record
    # }
    def on_message(msg)
      if msg.nil?
        # for future TCP heartbeat_request
        return
      end

      # TODO format error
      tag = msg[0].to_s
      entries = msg[1]

      if entries.class == String
        # PackedForward
        bytesize = tag.bytesize + entries.bytesize
        measure_time(:on_message, bytesize) do
          es = MessagePackEventStream.new(entries, @cached_unpacker)
          Engine.emit_stream(tag, es)
        end

      elsif entries.class == Array
        # Forward
        es = MultiEventStream.new
        entries.each {|e|
          record = e[1]
          next if record.nil?
          time = e[0].to_i
          time = (now ||= Engine.now) if time == 0
          es.add(time, record)
        }
        measure_time(:on_message, 0) do
          Engine.emit_stream(tag, es)
        end

      else
        # Message
        record = msg[2]
        return if record.nil?
        time = msg[1]
        time = Engine.now if time == 0
        bytesize = time.size + record.to_s.bytesize
        measure_time(:on_message, bytesize) do
          Engine.emit(tag, time, record)
        end
      end
    end

    def measure_time(caller, size)
      @elapsed ? @elapsed.measure_time(caller, size) { yield } : yield
    end

    class Handler < Coolio::Socket
      def initialize(io, linger_timeout, log, on_message, elapsed)
        super(io)
        if io.is_a?(TCPSocket)
          opt = [1, linger_timeout].pack('I!I!')  # { int l_onoff; int l_linger; }
          io.setsockopt(Socket::SOL_SOCKET, Socket::SO_LINGER, opt)
        end
        @on_message = on_message
        @log = log
        @log.trace {
          remote_port, remote_addr = *Socket.unpack_sockaddr_in(@_io.getpeername) rescue nil
          "accepted fluent socket from '#{remote_addr}:#{remote_port}': object_id=#{self.object_id}"
        }
        @elapsed = elapsed
      end

      def on_connect
      end

      def on_read(data)
        first = data[0]
        if first == '{' || first == '['
          m = method(:on_read_json)
          @y = Yajl::Parser.new
          @y.on_parse_complete = @on_message
        else
          m = method(:on_read_msgpack)
          @u = MessagePack::Unpacker.new
        end

        (class << self; self; end).module_eval do
          define_method(:on_read, m)
        end
        m.call(data)
      end

      def measure_time(caller, size)
        @elapsed ? @elapsed.measure_time(caller, size) { yield } : yield
      end

      def on_read_json(data)
        measure_time(:on_read, data.bytesize) do
          @y << data
        end
      rescue => e
        @log.error "forward error", :error => e, :error_class => e.class
        @log.error_backtrace
        close
      end

      def on_read_msgpack(data)
        measure_time(:on_read, data.bytesize) do
          @u.feed_each(data, &@on_message)
        end
      rescue => e
        @log.error "forward error", :error => e, :error_class => e.class
        @log.error_backtrace
        close
      end

      def on_close
        @log.trace { "closed fluent socket object_id=#{self.object_id}" }
      end
    end

    class HeartbeatRequestHandler < Coolio::IO
      def initialize(io, callback)
        super(io)
        @io = io
        @callback = callback
      end

      def on_readable
        begin
          msg, addr = @io.recvfrom(1024)
        rescue Errno::EAGAIN, Errno::EWOULDBLOCK, Errno::EINTR
          return
        end
        host = addr[3]
        port = addr[1]
        @callback.call(host, port, msg)
      rescue
        # TODO log?
      end
    end

    def on_heartbeat_request(host, port, msg)
      #log.trace "heartbeat request from #{host}:#{port}"
      begin
        @usock.send "\0", 0, host, port
      rescue Errno::EAGAIN, Errno::EWOULDBLOCK, Errno::EINTR
      end
    end
  end
end
